src/Kiota.Builder/Configuration/StructuredMimeTypesCollection.cs

```diff
@@ -0,0 +1,122 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace Kiota.Builder.Configuration;
+
+internal partial class StructuredMimeTypesCollection : ICollection<string>
+{
+    [GeneratedRegex(@"(?<mime>[^;]+);?q?=?(?<priority>[\d.]+)?", RegexOptions.IgnoreCase | RegexOptions.Compiled | RegexOptions.Singleline, 2000)]
+    private static partial Regex mimeTypesRegex();
+    private readonly static Regex mimeTypesRegexInstance = mimeTypesRegex();
+    private readonly Dictionary<string, float> _mimeTypes;
+
+    public int Count => _mimeTypes.Count;
+
+    public bool IsReadOnly => false;
+    public StructuredMimeTypesCollection() : this(Array.Empty<string>()) { }
+    public StructuredMimeTypesCollection(IEnumerable<string> mimeTypes)
+    {
+        ArgumentNullException.ThrowIfNull(mimeTypes);
+        _mimeTypes = mimeTypes.Select(GetKeyAndPriority)
+                                .OfType<KeyValuePair<string, float>>()
+                                .ToDictionary(static x => x.Key, static x => x.Value, StringComparer.OrdinalIgnoreCase);
+    }
+    private static KeyValuePair<string, float>? GetKeyAndPriority(string rawFormat)
+    {
+        if (string.IsNullOrEmpty(rawFormat))
+            return null;
+        var match = mimeTypesRegexInstance.Match(rawFormat);
+        if (match.Success)
+        {
+            var priority = match.Groups["priority"].Success && float.TryParse(match.Groups["priority"].Value, CultureInfo.InvariantCulture, out var resultPriority) ? resultPriority : 1;
+            return new KeyValuePair<string, float>(match.Groups["mime"].Value, priority);
+        }
+        else
+        {
+            return null;
+        }
+    }
+    public IEnumerator<string> GetEnumerator()
+    {
+        return _mimeTypes.OrderByDescending(static x => x.Value).Select(NormalizeMimeType).GetEnumerator();
+    }
+    IEnumerator IEnumerable.GetEnumerator()
+    {
+        return GetEnumerator();
+    }
+    public bool Contains(string mimeType)
+    {
+        if (string.IsNullOrEmpty(mimeType))
+            return false;
+        return _mimeTypes.ContainsKey(mimeType);
+    }
+    public float? GetPriority(string mimeType)
+    {
+        if (string.IsNullOrEmpty(mimeType))
+            return null;
+        return _mimeTypes.TryGetValue(mimeType, out var priority) ? priority : null;
+    }
+
+    public void Add(string item)
+    {
+        if (GetKeyAndPriority(item) is { } result)
+            _mimeTypes.TryAdd(result.Key, result.Value);
+    }
+
+    public void Clear()
+    {
+        _mimeTypes.Clear();
+    }
+
+    public void CopyTo(string[] array, int arrayIndex)
```

Comment:
It might be wise to document this method since it wasn't clear to me what its behavior was until I read the code.

Reply:
this is due to interface implementation, I don't believe it ends up actually being called in kiota's context. Added the inheritdoc comment

Reply:
I think you have to change the docs since your implementation performs a sort before copying so the resulting array is different from the source collection.

Reply:
Or I could remove the sorting, I don't think we actually need it. Thoughts?

Reply:
It doesn't have to do with the idempotency requirement? If not, removing it is fine. If so, I think the sorting should also be in the collection somehow. For example, having an extra sorted collection or using a sorted dictionary.

A related suggestion if you care about improving performance, Removing the `.ToArray()` so we use the already allocated array that has been passed in by the callers should reduce allocations and GC pressure.

Reply:
turns out I was wrong, this method is used when calling .ToList() on that collection. And yes we need to keep the ordering in so the kiota lock generation is idempotent.

As for the ToArray, I don't think there's an easy way around as:
- it gives us access to the array copy to method so we don't have to re-implement it.
- reimplementing the method would require to know the length and being able to index into the "collection"
- array is the smallest data structure that fullfills those requirements (as opposed to list)

Reply:
Implementing ToArray only needs iteration between the collection's begin and array.Length - arrayIndex. But it was just a suggestion. I feel there's a lot of unnecessary work happening in this method. For example, all the `Array.CopyTo` failure cases still create a new array that will be discarded when the exception is thrown. If it's not expected to cause an issue, I will resolve and leave the comment. 🙂